bdns/problem.go

```diff
@@ -9,30 +9,55 @@ import (
 	"fmt"
 	"net"
 
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/miekg/dns"
 	"github.com/letsencrypt/boulder/probs"
 )
 
-const detailDNSTimeout = "DNS query timed out"
-const detailDNSNetFailure = "DNS networking error"
-const detailServerFailure = "Server failure at resolver"
+type dnsError struct {
+	recordType uint16
+	hostname   string
+	// Exactly one of rCode or underlying should be set.
+	underlying error
+	rCode      int
+}
+
+func (d dnsError) Error() string {
+	var detail string
+	if d.underlying != nil {
+		if netErr, ok := d.underlying.(*net.OpError); ok {
+			if netErr.Timeout() {
+				detail = detailDNSTimeout
+			} else {
+				detail = detailDNSNetFailure
+			}
+		}
+	} else if d.rCode != dns.RcodeSuccess {
+		detail = dns.RcodeToString[d.rCode]
+	} else {
+		detail = detailServerFailure
+	}
+	return fmt.Sprintf("DNS problem: %s looking up %s for %s", detail,
+		dns.TypeToString[d.recordType], d.hostname)
+}
+
+const detailDNSTimeout = "query timed out"
+const detailDNSNetFailure = "networking error"
+const detailServerFailure = "server failure at resolver"
 
 // ProblemDetailsFromDNSError checks the error returned from Lookup...  methods
 // and tests if the error was an underlying net.OpError or an error caused by
 // resolver returning SERVFAIL or other invalid Rcodes and returns the relevant
 // core.ProblemDetails. The detail string will contain a mention of the DNS
 // record type and domain given.
-func ProblemDetailsFromDNSError(recordType, domain string, err error) *probs.ProblemDetails {
-	detail := detailServerFailure
-	if netErr, ok := err.(*net.OpError); ok {
-		if netErr.Timeout() {
-			detail = detailDNSTimeout
-		} else {
-			detail = detailDNSNetFailure
+func ProblemDetailsFromDNSError(err error) *probs.ProblemDetails {
```

Comment:
Doing it this way means that, unless the error is a dnsError, we lose the hostname and record type the error came from was for.

There are could be error types that can occur during look up. For instance, [context.DeadlineExceeded](https://godoc.org/golang.org/x/net/context#DeadlineExceeded) will start happening in the retry branch. Over time, it's possible that we'll add more. When those happen, we will have less helpful messages for our users.